src/toxcore/dht_new/dht_impl.rs

```diff
@@ -0,0 +1,259 @@
+
+        DhtPacket {
+            packet_kind: dp.kind(),
+            pk: *pk,
+            nonce: *nonce,
+            payload: payload,
+        }
+
+    }
+
+    /**
+    Decrypt payload and try to parse it as packet type.
+
+    To get info about it's packet type use
+    [`.kind()`](./struct.DhtPacket.html#method.kind) method.
+
+    Returns `None` in case of faliure:
+
+    - fails to decrypt
+    - fails to parse as given packet type
+    */
+    /* TODO: perhaps switch to using precomputed symmetric key?
+              - given that computing shared key is apparently the most
+                costly operation when it comes to crypto, using precomputed
+                key might (would significantly?) lower resource usage
+
+                Alternatively, another method `get_payloadnm()` which would use
+                symmetric key.
+    */
+    pub fn get_payload(&self, own_secret_key: &SecretKey) -> Result<Option<DhtPacketPayload>, Error>
+    {
+        debug!(target: "DhtPacket", "Getting packet data from DhtPacket.");
+        trace!(target: "DhtPacket", "With DhtPacket: {:?}", self);
+        let decrypted = match open(&self.payload, &self.nonce, &self.pk,
+                            own_secret_key) {
+            Ok(d) => d,
+            Err(_) => {
+                debug!("Decrypting DhtPacket failed!");
+                return Ok(None);
+            },
+        };
+
+        match self.packet_kind {
+            PacketKind::PingReq => {
+                if let IResult::Done(_, packet) = PingRequest::from_bytes(&decrypted) {
+                    Ok(Some(DhtPacketPayload::PingRequest(packet)))                    
+                } else {
+                    Ok(None)
+                }
+            },
+            PacketKind::PingResp => {
+                if let IResult::Done(_, packet) = PingResponse::from_bytes(&decrypted) {
+                    Ok(Some(DhtPacketPayload::PingResponse(packet)))
+                } else {
+                    Ok(None)
+                }
+            },
+            PacketKind::GetN => {
+                if let IResult::Done(_, packet) = GetNodes::from_bytes(&decrypted) {
+                    Ok(Some(DhtPacketPayload::GetNodes(packet)))
+                } else {
+                    Ok(None)
+                }
+            },
+            PacketKind::SendN => {
+                if let IResult::Done(_, packet) = SendNodes::from_bytes(&decrypted) {
+                    Ok(Some(DhtPacketPayload::SendNodes(packet)))
+                } else {
+                    Ok(None)
+                }
+            },
+            _ => Ok(None)
+        }
+    }
+
+    /**
+    Create DHT Packet with [`Ping`](./struct.Ping.html) response to `Ping`
+    request that packet contained.
+
+    Nonce for the response is automatically generated.
+    */
+    // Because UDP codec and tokio use DhtBase for send/receive packet, 
+    // this function returns DhtBase type object
+    pub fn ping_response(&self,
```

Comment:
I think this method should be located inside DHT Node but not inside DhtPacket

Reply:
I agree.
I moved this function to DhtPacket.rs